### PR TITLE
iox-1612 Move `std::string` dependency to `iceoryx_dust`

### LINKED ISSUE #1612: Replace `operator std::string()` with `toStdString()` method
## Brief feature description

Replace `operator std::string()` with either `toStdString()` method to make it more verbose or a `toStdString()` free function.

See:
```cpp
auto bla1 = myCxxString.toStdString(); // proposal 1
auto bla2 = std::string(myCxxString.c_str()); // old style, maybe sufficient
auto bla3 = toStdString(myCxxString); // proposal 2

auto bla4 = myCxxString; // will be forbidden
```

Another awesome approach would be to declare a free template function called `into` which can be specialized for any arbitrary types internally or used by users to perform conversions. This would be a way to support explicit conversions with free functions.

See:
```cpp
namespace iox {
// defined maybe in cxx/type_traits.hpp
template<typename Source, typename Destination>
Destination into(const Source& source);
}

// defined in an header
template<>
string<100> iox::into(const std::string& source);

// implemented in cpp file 
```

This approach would not allow us to convert arbitrary strings but we could use it for the most common string aliases like `IdString` etc. But if we have multiple string aliases with the same length we would get a ambiguity error in the linker.
If we put some brains into it we maybe solve the issue and would have a universal conversion idiom in iceoryx.

## Important note

This could also solve the issue we currently have in `cxx::convert` that were we are unable to partially overload `fromString` so that we could convert all `cxx::string`s.
At the moment the conversion has to be explicitly implemented for every string capacity.

Due to a lossy conversion with truncation to the capacity of the fixed string, it should be visible to the user that this might happen and the user must be given the choice to prevent the truncation.

This can be achieved by the following API
- [x] `iox::string<N> into<iox::lossy<iox::string<N>>>(std::string)`
- [ ] `iox::optional<iox::string<N>> into<iox::optional<iox::string<N>>>(std::string)`
    - commit [bf168d1](https://github.com/eclipse-iceoryx/iceoryx/pull/1889/commits/bf168d1242b41e4863a4b9f5a2dd73cef335939c) needs to be reverted and additional tests are required

Extracted from: #1196

### PR DESCRIPTION
**Pull request by mossmaurice**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer
* Move `std::string` dependencies to `iceoryx_dust`
   * Users have to include the support header in order to use the following functions with `std::string`
      * comparison operators
      * `operator<<`
      * `insert`
      * `find`
      * `unsafe_append` 

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #1612

### CODE REVIEW

File: iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp [hunk 1]
```diff
@@ -108,4 +110,618 @@ TYPED_TEST(StdString_test, StringOfSizeCapaToSTDStringConvResultsInSizeCapa)
     EXPECT_THAT(testString2.size(), Eq(STRINGCAP));
     EXPECT_THAT(testString2.c_str(), StrEq(testString1.substr(0, STRINGCAP)));
 }
+
+TYPED_TEST(StdString_test, UnsafeAppendEmptyStdStringWorks)
```
Comment (mossmaurice) [inline iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp:114]: @elBoberido @FerdinandSpitzschnueffler These tests are not new, just the `std::string` part separated from the other test cases.

File: iceoryx_hoofs/test/moduletests/test_vocabulary_string.cpp [hunk 2]
```diff
@@ -2939,48 +2706,6 @@ TEST(String100, FindFirstOfForNotIncludedStringLiteralFails)
     EXPECT_THAT(res.has_value(), Eq(false));
 }
 
-TEST(String100, FindFirstOfForSTDStringInNotEmptyStringWorks)
```
Comment (elBoberido) [inline iceoryx_hoofs/test/moduletests/test_vocabulary_string.cpp:2942]: @FerdinandSpitzschnueffler not for this PR but do we need the `String100` test and shouldn't these test cases already be covered by typed tests?

File: iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp [hunk 3]
```diff
@@ -108,4 +110,618 @@ TYPED_TEST(StdString_test, StringOfSizeCapaToSTDStringConvResultsInSizeCapa)
...
+    res = testString.find_first_of(testStdString2);
+    ASSERT_THAT(res.has_value(), Eq(true));
+    EXPECT_THAT(res.value(), Eq(0U));
+
+    res = testString.find_first_of(testStdString2, 1U);
+    ASSERT_THAT(res.has_value(), Eq(true));
+    EXPECT_THAT(res.value(), Eq(3U));
+}
+
+TEST(String100, FindFirstOfForNotIncludedSTDStringFails)
```
Comment (elBoberido) [inline iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp:623]: This is also something for a follow up together with `test_vocabulary_string.cpp`. It is weird to have typed tests intermixed with non typed tests

File: iceoryx_dust/include/iceoryx_dust/cxx/std_string_support.hpp [hunk 4]
```diff
@@ -19,10 +19,38 @@
...
+};
+
+template <>
+struct GetSize<std::string>
+{
+    static uint64_t call(const std::string& data) noexcept
+    {
+        return data.size();
+    }
+};
```
Comment (elBoberido) [inline iceoryx_dust/include/iceoryx_dust/cxx/std_string_support.hpp:51]: It seems three is currently no default implementation for `GetSize` and `GetData`. What do you think about adding a default implementation with a `static_assert(always_false_v<T>, "Please implement bla bla");`?

I'm not sure if something similar could be done for `is_custom_string` in order to have a sane error message on compilations.

**Review by elBoberido — COMMENTED** [review]
Looks good. Only a few questions regarding cryptic error messages when compiling

File: iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp [hunk 5]
```diff
@@ -108,4 +110,618 @@ TYPED_TEST(StdString_test, StringOfSizeCapaToSTDStringConvResultsInSizeCapa)
...
+    // NOLINTNEXTLINE(hicpp-avoid-c-arrays, cppcoreguidelines-avoid-c-arrays)
+    char testCharArray[STRINGCAP + 1U];
+    for (uint64_t i = 0U; i < STRINGCAP; ++i)
+    {
+        // NOLINTJUSTIFICATION no other way to populate testCharArray
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-constant-array-index)
+        testCharArray[i] = 'M';
+    }
+
+    const std::string testStdString = &testCharArray[0];
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp:292]: Seems like this was necessary for the "char array part" of the test and can be now replaced with  `const std::string testStdString(STRINGCAP, 'M');`

File: iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp [hunk 6]
```diff
@@ -108,4 +110,621 @@ TYPED_TEST(StdString_test, StringOfSizeCapaToSTDStringConvResultsInSizeCapa)
...
+    // NOLINTNEXTLINE(hicpp-avoid-c-arrays, cppcoreguidelines-avoid-c-arrays)
+    char testCharArray[STRINGCAP + 1U];
+    for (uint64_t i = 0U; i < STRINGCAP; ++i)
+    {
+        // NOLINTJUSTIFICATION no other way to populate testCharArray
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-constant-array-index)
+        testCharArray[i] = 'L';
+    }
+    testCharArray[STRINGCAP] = '\0';
+    const std::string testStdString = &testCharArray[0];
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp:314]: Similar here.

File: iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp [hunk 7]
```diff
@@ -108,4 +110,621 @@ TYPED_TEST(StdString_test, StringOfSizeCapaToSTDStringConvResultsInSizeCapa)
...
+}
+
+TYPED_TEST(StdString_test, CompareEqStringAndStdStringWithDifferentCapaResultsInZero)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "8a42e8e1-88c4-43bb-ae7f-1f49ddf3161b");
+    using MyString = typename TestFixture::stringType;
+    constexpr auto STRINGCAP = MyString::capacity();
+    std::string temp(STRINGCAP, 'M');
+    ASSERT_THAT(this->testSubject.unsafe_assign(temp.c_str()), Eq(true));
+
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp:335]:
```suggestion
```

File: iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp [hunk 8]
```diff
@@ -108,4 +110,621 @@ TYPED_TEST(StdString_test, StringOfSizeCapaToSTDStringConvResultsInSizeCapa)
...
+    char testCharArray[STRINGCAP + 1U];
+    for (uint64_t i = 0U; i < STRINGCAP; ++i)
+    {
+        // NOLINTJUSTIFICATION no other way to populate testCharArray
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-constant-array-index)
+        testCharArray[i] = 'L';
+    }
+    testCharArray[STRINGCAP] = '\0';
+
+    const std::string testStdString = &testCharArray[0];
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp:474]: Similar here.

File: iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp [hunk 9]
```diff
@@ -108,4 +110,621 @@ TYPED_TEST(StdString_test, StringOfSizeCapaToSTDStringConvResultsInSizeCapa)
...
+    using MyString = typename TestFixture::stringType;
+    constexpr auto STRINGCAP = MyString::capacity();
+    std::string testString(STRINGCAP + 6U, 'M');
+    string<STRINGCAP + 6U> sut;
+    ASSERT_THAT(sut.unsafe_assign(testString.c_str()), Eq(true));
+    std::string foo(STRINGCAP, 'M');
+    EXPECT_THAT(sut.compare(foo), Gt(0));
+}
+
+/// @note constexpr int64_t compare(const std::string& other) const noexcept
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp:227]: Please move this comment to line 215.

File: iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp [hunk 10]
```diff
@@ -108,4 +110,621 @@ TYPED_TEST(StdString_test, StringOfSizeCapaToSTDStringConvResultsInSizeCapa)
...
+
+    // append std::string
+    sut = "i";
+    sut.append(TruncateToCapacity, testStdString);
+    EXPECT_THAT(sut.capacity(), Eq(RESULT_CAPACITY));
+    EXPECT_THAT(sut.size(), Eq(7U));
+    EXPECT_THAT(std::memcmp(sut.c_str(), expectedString.c_str(), sut.size()), Eq(0));
+}
+
+TYPED_TEST(StdString_test, FindStdStringInEmptyStringFails)
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp:555]: What about `FindEmptyStdStringInEmptyStringWorks`?

File: iceoryx_hoofs/test/moduletests/test_vocabulary_string.cpp [hunk 11]
```diff
@@ -2782,29 +2575,11 @@ TEST(String100, FindNotIncludedStringLiteralFails)
...
-    ASSERT_THAT(res.has_value(), Eq(true));
-    EXPECT_THAT(res.value(), Eq(4U));
-}
-
 TEST(String100, FindNotIncludedSTDStringFails)
 {
     ::testing::Test::RecordProperty("TEST_ID", "8b2116c9-5f7d-48b4-8c26-cb3b71cf0ea2");
     string<100U> testString("Kernfusionsbaby");
-    std::string testStdString = "abc";
+    string<100U> testStdString = "abc";
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/test/moduletests/test_vocabulary_string.cpp:2582]: Please revert the change and move this test to test_cxx_string_conversion.cpp

File: iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp [hunk 12]
```diff
@@ -108,4 +110,621 @@ TYPED_TEST(StdString_test, StringOfSizeCapaToSTDStringConvResultsInSizeCapa)
...
+/// @note template <typename T>
+/// iox::cxx::optional<uint64_t> find_first_of(const T& t, uint64_t pos = 0) const noexcept
+TYPED_TEST(StdString_test, FindFirstOfFailsForEmptyStdStringInEmptyString)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "207671e4-cef3-40d2-8984-e8ae5c2b42ec");
+    using MyString = typename TestFixture::stringType;
+    constexpr auto STRINGCAP = MyString::capacity();
+    string<STRINGCAP + 5U> testString;
+    auto res = this->testSubject.find_first_of(testString);
+    EXPECT_THAT(res.has_value(), Eq(false));
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp:578]: This is not needed here.

File: iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp [hunk 13]
```diff
@@ -108,4 +110,621 @@ TYPED_TEST(StdString_test, StringOfSizeCapaToSTDStringConvResultsInSizeCapa)
...
+}
+
+TYPED_TEST(StdString_test, FindFirstOfForStdStringInEmptyStringFails)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "cfd4a842-64e3-4a2c-afc9-d98f93f1f8f4");
+    using MyString = typename TestFixture::stringType;
+    constexpr auto STRINGCAP = MyString::capacity();
+    string<STRINGCAP + 5U> testString("a");
+    auto res = this->testSubject.find_first_of(testString);
+    EXPECT_THAT(res.has_value(), Eq(false));
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp:592]: Same here.

File: iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp [hunk 14]
```diff
@@ -108,4 +110,621 @@ TYPED_TEST(StdString_test, StringOfSizeCapaToSTDStringConvResultsInSizeCapa)
...
+/// @note template <typename T>
+/// iox::cxx::optional<uint64_t> find_last_of(const T& t, uint64_t pos = 0) const noexcept
+TYPED_TEST(StdString_test, FindLastOfFailsForEmptyStdStringInEmptyString)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "15f72273-8b90-407f-b7d0-07372f3cee29");
+    using MyString = typename TestFixture::stringType;
+    constexpr auto STRINGCAP = MyString::capacity();
+    string<STRINGCAP + 5U> testString;
+    auto res = this->testSubject.find_last_of(testString);
+    EXPECT_THAT(res.has_value(), Eq(false));
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp:650]: Same here.

File: iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp [hunk 15]
```diff
@@ -108,4 +110,621 @@ TYPED_TEST(StdString_test, StringOfSizeCapaToSTDStringConvResultsInSizeCapa)
...
+}
+
+TYPED_TEST(StdString_test, FindLastOfForStdStringInEmptyStringFails)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "f8b48cdb-7165-41d6-9eb3-f148c6edd859");
+    using MyString = typename TestFixture::stringType;
+    constexpr auto STRINGCAP = MyString::capacity();
+    string<STRINGCAP + 5U> testString("a");
+    auto res = this->testSubject.find_last_of(testString);
+    EXPECT_THAT(res.has_value(), Eq(false));
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp:664]: Same here.

File: iceoryx_hoofs/test/moduletests/test_vocabulary_string.cpp (hunk 2 above)
Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/test/moduletests/test_vocabulary_string.cpp:2942]: The initial reason for the `String100` tests was that some tests require a `sut` capacity > 1, e.g. `UnsafeCharToStringConstrIncludingNullCharWithCountResultsInSizeCount`. We could rewrite the tests to typed tests but 3 years ago I thought the `String100` approach was a good idea...
To answer your question: I doubt that the `String100` tests are already covered by typed tests.

Comment (elBoberido) [reply iceoryx_hoofs/test/moduletests/test_vocabulary_string.cpp:2942]: Okay, then we can keep it but maybe rename them at some time. They remind me of `CString100` 😱

Comment (mossmaurice) [reply iceoryx_hoofs/test/moduletests/test_vocabulary_string.cpp:2942]: Something for a follow-up PR. Closing.

File: iceoryx_hoofs/test/moduletests/test_vocabulary_string.cpp [hunk 16]
```diff
@@ -84,7 +84,7 @@ TYPED_TEST(stringTyped_test, CopyConstructStringOfSizeCapaResultsInSizeCapa)
     string<STRINGCAP> fuu(this->testSubject);
     EXPECT_THAT(fuu.capacity(), Eq(STRINGCAP));
     EXPECT_THAT(fuu.size(), Eq(STRINGCAP));
-    EXPECT_THAT(fuu.c_str(), StrEq(testString));
+    EXPECT_THAT(fuu.c_str(), StrEq(testString.c_str()));
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_hoofs/test/moduletests/test_vocabulary_string.cpp]: Already fixed on the latest commit.

File: iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp (hunk 10 above)
Comment (mossmaurice) [reply iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp:555]: @FerdinandSpitzschnueffler Why `Empty`? The `std::string` isn't empty here.

File: iceoryx_hoofs/test/moduletests/test_vocabulary_string.cpp (hunk 11 above)
Comment (mossmaurice) [reply iceoryx_hoofs/test/moduletests/test_vocabulary_string.cpp:2582]: Good catch, done!

File: iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp (hunk 10 above)
Comment (FerdinandSpitzschnueffler) [reply iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp:555]: Yes, I meant that's a missing test case.

File: iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp (hunk 12 above)
Comment (mossmaurice) [reply iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp:578]: True, this is already in `FindFirstOfFailsForEmptyStringInEmptyString` with test id `21f90f13-6b15-4ce1-9258-c21154b6043c`.

File: iceoryx_hoofs/vocabulary/include/iox/string.hpp [hunk 17]
```diff
@@ -26,46 +26,53 @@
...
+{
+};
+
 // AXIVION DISABLE STYLE AutosarC++19_03-A18.1.1 : C-array type usage is intentional
 
 template <typename T, typename ReturnType>
 using IsStringOrCharArrayOrChar =
-    typename std::enable_if<((is_cxx_string<T>::value || is_char_array<T>::value)
-                             || (std::is_same<T, std::string>::value || std::is_same<T, char>::value)),
+    typename std::enable_if<(is_cxx_string<T>::value || is_char_array<T>::value || std::is_same<T, char>::value
```
Comment (elBoberido) [inline iceoryx_hoofs/vocabulary/include/iox/string.hpp:49]: Since you changed `IsCxxStringOrCharArray` to `IsIoxStringOrCharArray` would it make sense to change `is_cxx_string` to `is_iox_string`?

File: iceoryx_hoofs/vocabulary/include/iox/string.hpp [hunk 18]
```diff
@@ -380,12 +387,12 @@ class string final
...
     string& operator+=(const T&) noexcept;
 
-    /// @brief appends a iox::string/string literal/std::string to the end of this. If this' capacity is too
+    /// @brief appends a iox::string/string literal/custom string to the end of this. If this' capacity is too
     /// small for appending the whole string (literal), the remainder of the characters are truncated.
     ///
     /// @param [in] TruncateToCapacity_t is a compile time variable which is used to make the user aware of the possible
     /// truncation
-    /// @param [in] str is the iox::string/string literal/std::string to append
+    /// @param [in] str is the iox::string/string/custom string literal to append
```
Comment (elBoberido) [inline iceoryx_hoofs/vocabulary/include/iox/string.hpp:395]:
```suggestion
    /// @param [in] str is the iox::string/string literal/custom string to append
```

File: iceoryx_hoofs/vocabulary/include/iox/detail/string_internal.hpp [hunk 19]
```diff
@@ -67,9 +67,14 @@ struct GetCapa<char>
     static constexpr uint64_t capa{1U};
 };
 
-/// @brief struct to get size of iox::string/std::string/char array/char
+/// @brief struct to get size of iox::string/char array/char
```
Comment (elBoberido) [inline iceoryx_hoofs/vocabulary/include/iox/detail/string_internal.hpp:70]:
```suggestion
/// @brief struct to get size of iox::string/custom string/char array/char
```

File: iceoryx_hoofs/vocabulary/include/iox/detail/string_internal.hpp [hunk 20]
```diff
@@ -111,9 +107,14 @@ struct GetSize<char>
     }
 };
 
-/// @brief struct to get a pointer to the char array of the fixed string/string literal/std::string
+/// @brief struct to get a pointer to the char array of the fixed string/string literal
```
Comment (elBoberido) [inline iceoryx_hoofs/vocabulary/include/iox/detail/string_internal.hpp:110]:
```suggestion
/// @brief struct to get a pointer to the char array of the fixed string/string literal/custom string
```

File: iceoryx_hoofs/vocabulary/include/iox/string.hpp [hunk 21]
```diff
@@ -26,46 +26,53 @@
...
+template <typename T>
+struct is_custom_string : public std::false_type
+{
+};
+
 // AXIVION DISABLE STYLE AutosarC++19_03-A18.1.1 : C-array type usage is intentional
 
 template <typename T, typename ReturnType>
 using IsStringOrCharArrayOrChar =
-    typename std::enable_if<((is_cxx_string<T>::value || is_char_array<T>::value)
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_hoofs/vocabulary/include/iox/string.hpp:41]: Fixed.

File: iceoryx_hoofs/vocabulary/include/iox/detail/string_internal.hpp (hunk 19 above)
Comment (mossmaurice) [reply iceoryx_hoofs/vocabulary/include/iox/detail/string_internal.hpp:70]: @elBoberido I re-phrased it a bit.

File: iceoryx_hoofs/vocabulary/include/iox/detail/string_internal.hpp (hunk 20 above)
Comment (mossmaurice) [reply iceoryx_hoofs/vocabulary/include/iox/detail/string_internal.hpp:110]: @elBoberido I moved the doxygen description to or custom struct to `std_string_support.hpp`.

File: iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp [hunk 22]
```diff
@@ -108,4 +110,593 @@ TYPED_TEST(StdString_test, StringOfSizeCapaToSTDStringConvResultsInSizeCapa)
...
+    constexpr auto STRINGCAP = MyString::capacity();
+    constexpr auto RESULT_CAPACITY = STRINGCAP + 10U;
+    const std::string expectedString{"ice\0ryx", 7};
+
+    string<RESULT_CAPACITY> sut("i");
+    const string<RESULT_CAPACITY> testCxxString(TruncateToCapacity, expectedString.substr(1).c_str(), 6U);
+    const std::string testStdString = expectedString.substr(1);
+
+    // append std::string
+    sut = "i";
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp:515]: This is not needed because `sut` is already `i` after line 510.

File: iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp [hunk 23]
```diff
@@ -108,4 +110,593 @@ TYPED_TEST(StdString_test, StringOfSizeCapaToSTDStringConvResultsInSizeCapa)
...
+    // NOLINTJUSTIFICATION required to verify string literal functionality of iox::string
+    // NOLINTNEXTLINE(hicpp-avoid-c-arrays, cppcoreguidelines-avoid-c-arrays)
+    char testCharArray[TEST_CHAR_ARRAY_CAPACITY];
+    for (auto& c : testCharArray)
+    {
+        c = 'M';
+    }
+    testCharArray[STRINGCAP] = '\0';
+
+    const std::string testStdString = &testCharArray[0];
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_dust/test/moduletests/test_cxx_string_conversion.cpp:490]: This doesn't make sense for a std::string since `testStdString`'s capacity doesn't have to be greater than `STRINGCAP` here. This could be achieved using `reserve` but I'm not sure if this test brings much value. Maybe we can remove it.

File: iceoryx_hoofs/vocabulary/include/iox/detail/string_internal.hpp [hunk 24]
```diff
@@ -111,10 +110,16 @@ struct GetSize<char>
     }
 };
 
-/// @brief struct to get a pointer to the char array of the fixed string/string literal/std::string
+/// @brief generic empty implementation of the struct to get the data of a string
 template <typename T>
-struct GetData;
+struct GetData
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_hoofs/vocabulary/include/iox/detail/string_internal.hpp:115]: I'm not sure if that is a good idea. You mean to adding a `static_assert` to all of the types which inherit from `std::false_type`? E.g. in `iox/type_traits.hpp`. That would lead to the fact that "Substitution Failure Is Not An Error" (SFINAE) would become "Substitution Failure Is An Error", right? And thereby defeating the purpose.

File: doc/website/release-notes/iceoryx-unreleased.md [hunk 25]
```diff
@@ -1005,7 +1006,30 @@
...
+    if(myIoxString == myStdString)
+    {
+      ..
+    }
+
+    // after
+    #include "iceoryx_dust/cxx/std_string_support.hpp"
+
+    std::string myStdString("foo");
+    iox::string<3> myIoxString(TruncateToCapacity, myStdString);
```
Comment (FerdinandSpitzschnueffler) [inline doc/website/release-notes/iceoryx-unreleased.md:1025]: This should be now something like `iox::string<3> myIoxString(into<lossy<iox::string<3>>>(myStdString);`

Comment (mossmaurice) [reply doc/website/release-notes/iceoryx-unreleased.md:1025]: @FerdinandSpitzschnueffler Good catch, fixed.

File: iceoryx_hoofs/vocabulary/include/iox/string.hpp [hunk 26]
```diff
@@ -26,50 +26,57 @@
 #include <algorithm>
 #include <cstdint>
 #include <cstring>
-#include <iostream>
```
Comment (elBoberido) [inline iceoryx_hoofs/vocabulary/include/iox/string.hpp:29]: :partying_face:

**Review by elBoberido — APPROVED** [review]
Looks good

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

**Review by elBoberido — APPROVED** [review]
